### PR TITLE
Add Batch jacobi device kernels

### PR DESCRIPTION
**Pull request by pratikvn**
This PR adds the device kernels for batch Jacobi preconditioner.

### CODE REVIEW

**Review by MarcelKoch — DISMISSED** [review]
Overall looks good. Mostly minor comments regarding consistency left.

One thing I've noticed is that some of the functions in the device preconditioner classes explicitly synchronize at the end, while others don't. Maybe it would be best to find a consistent behavior. I think I would prefer to not synchronize in the functions, but instead require synchronization after the function call.

File: dpcpp/preconditioner/batch_preconditioners.hpp
Comment (pratikvn) [reply dpcpp/preconditioner/batch_preconditioners.hpp]: should not have been necessary in cuda/hip either. removed now

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (pratikvn) [reply common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:118]: choose pivot is meant to compute the pivot value, so I think it is natural to return the pivot there. Here `perm` is being updated, once the block has been inverted. Also, I think a `invert_dense_block` function returning a `perm` seems unintuitive.

File: common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc
Comment (pratikvn) [reply common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc:90]: Yes, I guess it is a question of where you do the sync. One would hope if there are two consecutive calls `__syncthreads()` one of them is optimized out. So, I have added a `__syncthreads()` here.

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (MarcelKoch) [reply common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:118]: It's returning an int anyway, so I don't understand the argument. But it's not really important.

File: test/preconditioner/CMakeLists.txt
Comment (MarcelKoch) [inline test/preconditioner/CMakeLists.txt:1]: nit
```suggestion
ginkgo_create_common_test(batch_jacobi_kernels)
```

**Review by MarcelKoch — APPROVED** [review]

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (yhmtsai) [inline common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:24]:
```suggestion
    const auto tid = threadIdx.x + blockIdx.x * blockDim.x;
```

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (yhmtsai) [inline common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:10]:
```suggestion
    const auto tid = threadIdx.x + blockIdx.x * blockDim.x;
```

File: common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc
Comment (yhmtsai) [inline common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc:118]: you can use safe_divide here

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (yhmtsai) [inline common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:99]:
```suggestion
    #pragma unroll
    for (int a = subwarp_grp.size() / 2; a > 0; a /= 2) {
```

File: dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (yhmtsai) [inline dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc:100]: different from cuda/hip

File: dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (yhmtsai) [inline dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc:209]: Maybe I missed it, but I do not find the corresponding function in cuda/hip

File: dpcpp/preconditioner/batch_scalar_jacobi.hpp.inc
Comment (yhmtsai) [inline dpcpp/preconditioner/batch_scalar_jacobi.hpp.inc:120]: safe_divide

File: dpcpp/preconditioner/batch_block_jacobi.hpp.inc
Comment (yhmtsai) [inline dpcpp/preconditioner/batch_block_jacobi.hpp.inc:49]: different sync place from cuda hip?

File: common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc
Comment (pratikvn) [reply common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc:118]: Safe divide sets the default to zero, but I want to set it to one

File: dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc
Comment (pratikvn) [reply dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc:209]: Yes, you are right. This is actually not needed.

File: test/preconditioner/batch_jacobi_kernels.cpp
Comment (pratikvn) [reply test/preconditioner/batch_jacobi_kernels.cpp:252]: Yes, it can still be somehow equal.

File: dpcpp/preconditioner/batch_block_jacobi.hpp.inc
Comment (pratikvn) [reply dpcpp/preconditioner/batch_block_jacobi.hpp.inc:49]: I guess it is not really necessary, but I think it should not hurt.

File: common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc
Comment (yhmtsai) [reply common/cuda_hip/preconditioner/batch_scalar_jacobi.hpp.inc:118]: ah, you are right

File: dpcpp/preconditioner/batch_block_jacobi.hpp.inc
Comment (yhmtsai) [reply dpcpp/preconditioner/batch_block_jacobi.hpp.inc:49]: the barrier should take quite some overhead.
If it is unnecessary, maybe remove it.

File: test/preconditioner/batch_jacobi_kernels.cpp
Comment (yhmtsai) [reply test/preconditioner/batch_jacobi_kernels.cpp:252]: If it possbly equal, maybe use `EXPECT_LE` not `EXPECT_LT`?

**Review by yhmtsai — CHANGES_REQUESTED** [review]
one major thing left: dealing with `d` value in `invert_dense_block` are different between sycl and cuda.

File: dpcpp/preconditioner/batch_block_jacobi.hpp.inc
Comment (pratikvn) [reply dpcpp/preconditioner/batch_block_jacobi.hpp.inc:49]: I would prefer to keep it for now. Will look into removing the unnecessary syncs at a later point.

**Review by yhmtsai — APPROVED** [review]

**Comment by pratikvn** [comment]
Will merge this only once #1396 has been merged.

**Comment by tcojean** [comment]
@pratikvn I will check if I can find the issue with the SYCL backend

**Pull request merged**
